pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -38,18 +38,110 @@ var (
 )
 
 // A stateful listener builder
+// Support the below intentions
+// 1. Use separate inbound capture listener(:15006) and outbound capture listener(:15001)
+// 2. The above listeners use bind_to_port sub listeners or filter chains.
 type ListenerBuilder struct {
 	node                   *model.Proxy
 	gatewayListeners       []*xdsapi.Listener
 	inboundListeners       []*xdsapi.Listener
 	outboundListeners      []*xdsapi.Listener
 	virtualListener        *xdsapi.Listener
 	virtualInboundListener *xdsapi.Listener
+	useInboundFilterChain  bool
+}
+
+func insertOriginalListenerName(chain *listener.FilterChain, listenerName string) {
+	if chain.Metadata == nil {
+		chain.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*types.Struct{},
+		}
+	}
+	if chain.Metadata.FilterMetadata[PilotMetaKey] == nil {
+		chain.Metadata.FilterMetadata[PilotMetaKey] = &types.Struct{
+			Fields: map[string]*types.Value{},
+		}
+	}
+	chain.Metadata.FilterMetadata[PilotMetaKey].Fields["original_listener_name"] =
+		&types.Value{Kind: &types.Value_StringValue{StringValue: listenerName}}
+}
+
+// Setup the filter chain match so that the match should work under both
+// - bind_to_port == false listener
```

Comment:
The transition will take more longer than I expected. 
There are a few tests depends on the implementation instead of the interface.
E.g. I fixed a permissive integration test in the most recent diff. It was failing even I haven't remove the bind listeners.
The rest of the week I prefer to consolidate the virtual case, including debugability and rollback flags. 
IMHO the best time to fully abandoning the legacy not-bind listeners is 1.4

Reply:
I think the rule is we can't have alpha features as default, and this is brand new ( in 1.2 it was mostly broken, we just tested that capture worked )